### DISCUSSION TITLE
InitializeComponent() is not recognized in brand new Avalonia 11.0.4 projects

### DISCUSSION THREAD

**Discussion by dprevato** [AvaloniaUI/Avalonia#13035] [Q&A]
**Describe the bug**
I just started a new project, using Avalonia, but I have the error "CS0103: The name InitializeComponent does not exist in the current context", in MainWindow.axaml.cs, without any code added from my part. All the required DLLs are present, I'm using Community.Toolkit.Mvvm v8.2.1. No idea about how to resolve the problem. Someone can help me?
**To Reproduce**

1. Create a new project, using the following settings:
- template "Avalonia .NET Core MVVM App"
- Framework ".NET 7.0"
- MVVM Toolkit "Community Toolkit", Avalonia Version "11.0.0-rc1"
- use compiled bindings
2. Build the brand new project; I see the preview of the MainWindow.
3. Open the file "MainWindow.xaml.cs: there is the error "Cannot resolve symbol 'InitializeComponent'". The message appears only as tooltip when I move the mouse cursor over the function call, but _no error is listed in the Error List window_.
4. I tried to update the nuget packages to the version 11.0.4, with no avail.

**Expected behavior**
I expect to have the method InitializeComponent resolved, so I can start to write my own code.

**Desktop (please complete the following information):**

- OS: Windows 11 Professional;
- Version 11.0.0-rc1 AND 11.0.4

**Comment by timunie** [comment]
rc1 is pretty outdated. Please update your templates to use latest version.

**Comment by timunie** [comment]
Let's make it a Q&A for now, as it's not reproducable using latest templates. Please upgrade and try again. If it still fails, file a minimal sample. Thx 🙏

**Reply by dprevato** [reply]
It's not clear to me how can I upgrade the Avalonia templates... can you help, please?

**Reply by thevortexcloud** [reply]
Run `dotnet new update` in a terminal/cmd window.

**Reply by dprevato** [reply]
Thank you for the answer; I did it, and Avalonia templates got updated to version 11.0.2 ; alas, the problem  is still there, "Cannot resolve symbol 'InitializeComponent'", exactly as I described in my post.

**Reply by thevortexcloud** [reply]
Did you recreate the project? All that does is update the template for new projects. It wont update existing projects.

On a side note, older versions of Avalonia (<11.0) require the Xaml generator package to generate code behind files properly. You should not need it for 11.0.

**Reply by dprevato** [reply]
I created a new project, initially with version 11.0.2 (the one of the updated templates), then after updating the packages to 11.0.4, but the problem is still there. 
I applied a workaround, a very ugly one, that I saw in the #9846 issue: I wrote myself the method InitializeComponent, making all the stuff work; but I think that the right place for such method is the parent class, Window.

![image](https://github.com/AvaloniaUI/Avalonia/assets/13785877/a07c28b6-8056-4897-8d99-455882745352)

**Reply by thevortexcloud** [reply]
I just tried this myself without any problems. Are you sure you are seeing a real build error? Visual Studio does sometimes like to say things are errors when they are not. Also try it from the command line to make sure it's not your IDE doing weird things:

`dotnet new avalonia.app`
`dotnet run`

That should make a new project and run it. If that does not work, also try running `dotnet restore` as the packages may be failing to download for some reason.

>  such method is the parent class, Window.

That's not how Avalonia works. Avalonia makes heavy use of source generators (which is why it's a partial class). There is also no requirement to use XAML at all with Avalonia. You can technically make all the controls in code (there is a third party project to do exactly that). That said It should automagically generate that method for you if your XAML is valid. If it's not valid you may run into non obvious problems like this.

**Reply by timunie** [reply]
@dprevato please attach the new project that doesn't work for you here for us to inspect.

**Reply by dprevato** [reply]
I created two projects, one using the command dotnet new avalonia.app, and the second using Visual Studio and the template "Avalonia .NET App" (that, I presume, is the same template). As you can see, the project created in command line works perfectly, the code generator produced the code for the method InitializeComponent. In the app created in Visual Studio, I have the error. The projects in the archive are exactly as they where created, I've not made any build, so you can test. 
[AvaloniaTests.zip](https://github.com/AvaloniaUI/Avalonia/files/12747178/AvaloniaTests.zip)

I have to stress that there is no code of mine, no XAML, not a single bit. One template works, the other not.

**Reply by timunie** [reply]
🤔 
I took your sample, opened it and just hit [F5], nothing more. I get: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/95688c12-e043-4d56-b511-9515b55b8333)

So I see no issue on my side. Files are generated as desired, see left tree. I'm out of ideas now. Sometimes VS is blocking source gen, as a file was marked as locked. You can try to clean generated files by forcing a deletion. See: 
- https://github.com/dotnet/roslyn/issues/48083
- https://learn.microsoft.com/en-us/answers/questions/1184090/looking-for-assistance-clearing-the-cache-for-upda
-

**Reply by thevortexcloud** [reply]
Certainly sounds like a Visual Studio problem. Which would not surprise me. Visual Studio does have a habit of breaking in weird subtle ways.

You can try compiling it from the command line too to see if that can get around it as a work around for the moment.

**Reply by Erik-Egnatosyan** [reply]
I have the same issue. I'm using Rider and newest version of avalonia and .net 9 (i try on .net 8 and still doesn't work). I'm updating everything that's is possible and nothing has changed. I deleted the file bin, obj, temp etc. The problem is still here and seems its not the problem of IDE.

**Reply by Myriad4Link** [reply]
To what I found (I used Rider), simply deleting "partial" statement in front of MainWindow and then add it back fix the problem.
tbh, wtf?

**Comment by dprevato** [comment]
I think exactly the same, now. I'll follow your advice, and create my
templates from the command line. Thank you for the assistance :)


Il giorno gio 28 set 2023 alle ore 11:13 thevortexcloud <
***@***.***> ha scritto:

**Reply by dprevato** [reply]
OK, I made some other tries, and I found that, with dotnet new, the template avalonia.app works fine, but the avalonia.mvvm one creates the error, like when I create the solution in Visual Studio. Pity. But this make me think that Visual Studio is not the one to blame, after all

**Reply by timunie** [reply]
@dprevato do you have the exact build number of your VS at hand probably? 

/cc @Takoooooo if you have an idea.

**Reply by dprevato** [reply]
Yep. The Visual Studio Enterprise version is 17.7.4, build number 17.7.34031.279

**Comment by dprevato** [comment]
I'm continuing to study the problem, and I noticed that the InitializeComponent method is defined in the Core.Views.MainWindow.g.cs file, and not only that: if I hover the mouse over the call that doesn't resolve, here is the definition and the description, as they are found in the other self-generated partial class. At runtime the error disappears, therefore you can live without solving it; but, forgive me, I feel as if I were in front of a splendid, perfectly white wall (Avalonia), with a speck in the corner: no matter how much effort I make, the eye continually falls on the speck :)

**Reply by thevortexcloud** [reply]
So to be clear it builds fine? Because if it does I'm back to thinking it's a Visual Studio problem.

**Reply by dprevato** [reply]
It builds fine and execute fine; I also think that it is some problem in Visual Studio, which does not connect the two partial MainWindow classes at design-time but does so at runtime; but we know well that the most fascinating part of the job is solving bugs...

**Reply by idea-voyager** [reply]
Try removing a reference to the XamlNameReferenceGenerator from your project.

**Comment by zeke202207** [comment]
I also encountered the same problem. Is there a solution?thanks @dprevato

**Reply by workgroupengineering** [reply]
You've probably mixed up multiple different versions of avalonia. Open your csproj and check the Avalonia `PackageReference` versions. If you use third-party components, also check which version of Avalonia they depend on.

**Reply by thevortexcloud** [reply]
As mentioned previously the problem can be caused by several things. That said the most likely cause for existing apps is invalid XAML. It all gets fed into a source generator and if it can't be parsed the source generator will fail, resulting in that method not existing.

**Reply by dprevato** [reply]
As I said above, the workaround I found is simply ignore the error, as the project compiles and runs without problems. I know, it's not a solution, but one cannot remain forever stuck, if he can avoid it... :)

**Reply by zeke202207** [reply]
You are right, maybe I have a little OCD, this problem will always be a pimple in the heart. Thank you  @dprevato , Thank you all!

**Reply by thevortexcloud** [reply]
The tl;dr for this thread is don't trust Visual Studio with build errors. It does weird things and it often lies. Always check the actual build output.

**Comment by chenjing1294** [comment]
@dprevato 
![33BJEAl8S8XCyK1j4iLwwA](https://github.com/user-attachments/assets/6e88ad0d-a58c-4ffe-922c-f23ff0396f26)

**Reply by unitrix0** [reply]
Helped me, also needed a rider restart

**Comment by AnnDevNet** [comment]
A possible track, unless the correction comes from another of my modifications :-)

I had the same problem in Debug but not in Release so I deleted this line from the project
`<PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.2.1" />`
I rebuild everything in debug (and in release) and no more errors.
I put the line back in place for the Avalonia.Diagnostics package and rebuild everything, and still no errors.
I also checked that adding an x:Name in the xaml added a reference to the auto-generated .g.cs file, when saving the xaml.

**Comment by MAC293** [comment]
Hi guys!
Are you still encountering this issue since the last time?
Mine is a little bit different, I get a "Cannot resolve symbol InitializeComponent()". I'm on MVVM AvaloniaUI on VS2022 using NET 8.

![InitializeComponent() Error](https://github.com/user-attachments/assets/5290787b-2738-48c5-9fb1-ea60669acd3e)

It's pretty odd when I change the target framework to 9, the problem disappears, but if I reboot VS, the issue comes back. It's a temporary fix. I don't get it. Even though, it runs with no problems. However, when looking for a method of any control it becomes such a tedious task, the autocomplete is dead cause of this.

Is there a definitive solution?

Regards

**Reply by timunie** [reply]
this comes from the source generators most likely. Nothing we can do here I guess.

**Reply by MAC293** [reply]
What a bummer. I've seen guys having it working correctly. It all comes down to their configuration. It's so damn annoying.

Thanks, dude.

**Reply by timunie** [reply]
You may want to give Rider a try.

**Reply by thevortexcloud** [reply]
Source generator support in most IDEs is still pretty terrible (especially when it comes to writing them), including in Visual Studio. Anyway, sometimes doing a clean and then deleting the `obj` folder _can_ help fix various problems. You may need to delete it multiple times though.

**Reply by MAC293** [reply]
I did all of that several times. The only "fix" is to change the target framework to 9.0, but it will revert after restarting VS.

**Reply by raphaelschmitz00** [reply]
It's been a while, but it's also still happening, so for anybody still looking this up:

What helped for me, in Rider, was to not use the xml namespace for the class name in the .axaml.
So instead of
```xml
xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
xmlns:myNamespace="clr-namespace:My.Namespace"
x:Class="myNamespace:MyClass"
```
I switched it to
```
xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
x:Class="My.Namespace.MyClass"
```
...and that seems to work better.

If I rename (or otherwise refactor) the view via IDE functionality,
I still tend to get this error, but it consistently works again if i do
clean project => unload project => wait for IDE to finish scanning => load project => build project

**Comment by Meta40k** [comment]
Ubuntu 24.04, dotnet version 9.0.102, Rider

dotnet new avalonia.app -n GuiCam
cd GuiCam
dotnet new avalonia.window -n CustomMessageBox -o component/CustomMessageBox

_I use a translator._ 
I created a new project and added a new component. I didn't do anything else and immediately get the error Cannot resolve symbol 'InitializeComponent'

why ?!

**Reply by thevortexcloud** [reply]
Please run

`dotnet restore`
`dotnet build`

If that does not work, run:
`dotnet new update`

Then generate a new project.

**Reply by Meta40k** [reply]
> Please run
> 
> `dotnet restore` `dotnet build`
> 
> If that does not work, run: `dotnet new update`
> 
> Then generate a new project.

I've tried a lot and I can't fix this situation. As @dprevato  wrote above, this does not affect my work and everything works for me. It's just that Reader highlights this method in red in all files.

**Reply by rabbitism** [reply]
> I've tried a lot and I can't fix this situation. As @dprevato  wrote above, this does not affect my work and everything works for me. It's just that Reader highlights this method in red in all files.

most of time you can just close ide, delete obj and bin folder, and restart ide.

**Reply by SteoTheFrog** [reply]
I get this on Linux with 11.2.3, but only on Linux. Switching to 11.0.0 makes it go away. It would be ideal it if could work with the latest version without red all over VS Code. Everything from every axaml file is not recognised by the IDE.

EDIT: I had to regenerate a new template and move everything from my old project to there, then it stopped happening. It was only affecting Linux though.

**Comment by corvinsz** [comment]
For me the problem was that when I added a new for example UserControl, it automatically added the following `ItemGroup`s to my project file.
Removing the `ItemGroup`s solved this problem.
![grafik](https://github.com/user-attachments/assets/e21c8082-32bd-4c16-b41b-1312d1ea9e04)

**Comment by mvinicius2k** [comment]
Same here. Change the **target framework** to **net8.0** solves this (for me). Avalonia packages only mention .NET 8 dependencies, maybe .NET 9 is not working now.

![image](https://github.com/user-attachments/assets/0215b6cf-5d2a-4110-b183-219b95d6b77e)

**Reply by timunie** [reply]
Net 9 works with latest Avalonia. Make sure to update to 11.2.6 or 11.3 beta

**Reply by arttsai** [reply]
I use Net 9 with Avalonia 11.2.7 and get the same error.  It doesn't know InitializeComponent(). 
I follow @mvinicius2k  suggestion clean the solution and switch to Net 8.  It works.

**Reply by levinit** [reply]
but my dotnet must be 6.0🥹

**Comment by kekekeks** [comment]
Check build logs, Roslyn reports errors as warnings by default.

**Reply by AndyE0** [reply]
Just started with Avalonia and hit this issue. It disappeared once on a rebuild but now is just there all the time. Cleaning, switching .NET versions between 8.0 and 9.0 makes no difference (I was on 8.0 anyway) and I'm on Avalonia 11.3.0. F12 finds the auto-generated code so it's more annoying than anything.

**Reply by MAC293** [reply]
> Just started with Avalonia and hit this issue. It disappeared once on a rebuild but now is just there all the time. Cleaning, switching .NET versions between 8.0 and 9.0 makes no difference (I was on 8.0 anyway) and I'm on Avalonia 11.3.0. F12 finds the auto-generated code so it's more annoying than anything.

Same here, dude! I just started an MVVM AvaloniaUI project from scratch, and still, I don't find a solution for this annoying bug/issue, and I did all those changes too, with no fix whatsoever.

**Comment by levinit** [comment]
add these lines in your behind code file:
```c#
using Avalonia.Markup.Xaml;


#... add the following lines after the constructor function
    private void InitializeComponent(){
        AvaloniaXamlLoader.Load(this);
    }
```

**Reply by sharpchen** [reply]
It can't be manually fixed like this. Control fields added by `x:Name` need to be initialized within it too, it may contain more details that should leave to be auto-generated

**Comment by MAC293** [comment]
<img width="330" height="46" alt="2026-01-27_21-55-42" src="https://github.com/user-attachments/assets/f0db3f79-0542-4230-95d9-458658631a6a" />

It's impossible to get rid of this bug/error/problem/etc. I've tried everything within my reach. Regardless of being unable to fix it, it still works. It's so weird.

**Reply by maxkatz6** [reply]
`Rider` is known to have issues with source generators sometimes.
Either way I would recommend reporting this issue to JetBrains youtrack.

Or, if you don't use Rider, - Visual Studio support (I just assumed from the color palette, and that Rider typically fails on that).

**Reply by maxkatz6** [reply]
For me, error disappears after app is built and run once.

**Reply by MAC293** [reply]
> `Rider` is known to have issues with source generators sometimes. Either way I would recommend reporting this issue to JetBrains youtrack.
> 
> Or, if you don't use Rider, - Visual Studio support (I just assumed from the color palette, and that Rider typically fails on that).

I'm using Resharper on VS Community 2022.

**Comment by JCH2k** [comment]
It's all in the project file.

Remove the <AdditionalFiles> for all XAML files and make sure that every XAML file has this (in one or more ItemsGroups):

```
<Compile Update="Path\To\ScanningView.axaml.cs">
      <DependentUpon>ScanningView.axaml</DependentUpon>
      <SubType>Code</SubType>
</Compile>
```

Then clear Rider caches and restart --> gone!
